### PR TITLE
[WIP] Feature/metrics

### PR DESCRIPTION
**Pull request by Mpdreamz**
This PR has a couple of additions 

`.AuditRequests()`

Control whether auditing as a whole should be enabled or not. The default is still `true`, will post BenchmarkDotNet  numbers to see if this setting is worth keeping in terms of allocations.

`.ProfileRequests()`

This will enable `Stopwatch.ElapsedTicks` backed tracked of elapsed time of audits and exposes a way for anything that has access to `RequestData` to attach `ClientProfileMeasurements`

`ScratchPad project`

I will cleanup `Program.cs` but i got fed up with hooking up linqpad or a random project to our local build output given that i have `1.x` `2.x` `5.x.` and master all checked out separately multiplied by different TFM's. 

`DebugInformation`

Elapsed either uses the DateTime Started/Ended properties or the stopwatch ticks depending on which is available. 

### Debug Information : Audit: No Profile: No

Setting `ProfileRequests` to true would have the same effect.

```
Valid NEST response built from a successful low level call on POST: /_all/object/_search
# Request:
<Request stream not captured or already read to completion by serializer. Set DisableDirectStreaming() on ConnectionSettings to force it to be set on t
he response.>
# Response:
<Response stream not captured or already read to completion by serializer. Set DisableDirectStreaming() on ConnectionSettings to force it to be set on
the response.>
```

### Debug Information : Audit: Yes Profile: No

```
Valid NEST response built from a successful low level call on POST: /_all/object/_search
# Audit trail of this API call (Total Elapsed: 00:00:06.3387517):
 - [1] PingFailure: Node: http://localhost:9209/ Exception: PipelineException Elapsed: 00:00:02.0334988
 - [2] PingFailure: Node: http://localhost:9208/ Exception: PipelineException Elapsed: 00:00:02.0034118
 - [3] PingFailure: Node: http://localhost:9201/ Exception: PipelineException Elapsed: 00:00:02.0020515
 - [4] PingSuccess: Node: http://localhost:9200/ Elapsed: 00:00:00.0150432
 - [5] HealthyResponse: Node: http://localhost:9200/ Elapsed: 00:00:00.2847464

.. continued...
```

### Debug Information : Audit: Yes Profile: Yes

```
Valid NEST response built from a successful low level call on POST: /_all/object/_search
# Audit trail of this API call (Total Elapsed: 00:00:06.3458422):
 - [1] PingFailure: Node: http://localhost:9209/ Exception: PipelineException Elapsed: 00:00:02.0382265
 --| Trace: [CreateHttpWebRequest] Elapsed: 00:00:00.0217832
 --| Trace: [GetResponse] Elapsed: 00:00:02.0108128

 - [2] PingFailure: Node: http://localhost:9208/ Exception: PipelineException Elapsed: 00:00:02.0027038
 --| Trace: [CreateHttpWebRequest] Elapsed: 00:00:00.0000738
 --| Trace: [GetResponse] Elapsed: 00:00:02.0025449

 - [3] PingFailure: Node: http://localhost:9201/ Exception: PipelineException Elapsed: 00:00:02.0026623
 --| Trace: [CreateHttpWebRequest] Elapsed: 00:00:00.0000422
 --| Trace: [GetResponse] Elapsed: 00:00:02.0025852

 - [4] PingSuccess: Node: http://localhost:9200/ Elapsed: 00:00:00.0149258
 --| Trace: [CreateHttpWebRequest] Elapsed: 00:00:00.0000296
 --| Trace: [GetResponse] Elapsed: 00:00:00.0114733
 --| Trace: [GetResponseStream] Elapsed: 00:00:00.0001276
 --| Trace: [SetSpecialTypes] Elapsed: 00:00:00.0001536

 - [5] HealthyResponse: Node: http://localhost:9200/ Elapsed: 00:00:00.2873238
 --| Trace: [CreateHttpWebRequest] Elapsed: 00:00:00.0000355
 --| Trace: [GetRequestStream] Elapsed: 00:00:00.1848662
 --| Trace: [Write] Elapsed: 00:00:00.1847828
 --| Trace: [GetResponse] Elapsed: 00:00:00.0019322
 --| Trace: [GetResponseStream] Elapsed: 00:00:00.0000011
 --| Trace: [Deserialize] Elapsed: 00:00:00.1000607

...continued..
```

and on .NET core 

```
Valid NEST response built from a successful low level call on POST: /_all/object/_search
# Audit trail of this API call (Total Elapsed: 00:00:10.4292999):
 - [1] PingFailure: Node: http://localhost:9209/ Exception: PipelineException Elapsed: 00:00:02.0973715
 --| Trace: [GetClient] Elapsed: 00:00:00.0256809
 --| Trace: [CreateHttpRequestMessage] Elapsed: 00:00:00.0045234
 --| Trace: [SendAsync blocking] Elapsed: 00:00:02.0590932

 - [2] PingFailure: Node: http://localhost:9208/ Exception: PipelineException Elapsed: 00:00:04.0060602
 --| Trace: [GetClient] Elapsed: 00:00:00.0000082
 --| Trace: [CreateHttpRequestMessage] Elapsed: 00:00:00.0000434
 --| Trace: [SendAsync blocking] Elapsed: 00:00:04.0058860

 - [3] PingFailure: Node: http://localhost:9201/ Exception: PipelineException Elapsed: 00:00:04.0149246
 --| Trace: [GetClient] Elapsed: 00:00:00.0000019
 --| Trace: [CreateHttpRequestMessage] Elapsed: 00:00:00.0000272
 --| Trace: [SendAsync blocking] Elapsed: 00:00:04.0148404

 - [4] PingSuccess: Node: http://localhost:9200/ Elapsed: 00:00:00.0137354
 --| Trace: [GetClient] Elapsed: 00:00:00.0000015
 --| Trace: [CreateHttpRequestMessage] Elapsed: 00:00:00.0000284
 --| Trace: [SendAsync blocking] Elapsed: 00:00:00.0098781
 --| Trace: [ReadAsStreamAsync blocking] Elapsed: 00:00:00.0000734
 --| Trace: [SetSpecialTypes] Elapsed: 00:00:00.0003089

 - [5] HealthyResponse: Node: http://localhost:9200/ Elapsed: 00:00:00.2972082
 --| Trace: [GetClient] Elapsed: 00:00:00.0000015
 --| Trace: [CreateHttpRequestMessage] Elapsed: 00:00:00.2158220
 --| Trace: [SendAsync blocking] Elapsed: 00:00:00.0045187
 --| Trace: [ReadAsStreamAsync blocking] Elapsed: 00:00:00.0000165
 --| Trace: [Deserialize] Elapsed: 00:00:00.0763005
````

### CODE REVIEW

File: src/Elasticsearch.Net/Auditing/Auditable.cs [hunk 1]
```diff
@@ -1,30 +1,50 @@
...
-		readonly IDateTimeProvider _dateTimeProvider;
+		public static Auditable Noop { get; } = new Auditable();
+
+		private readonly IDateTimeProvider _dateTimeProvider;
+		private readonly Func<long?> _getElapsed;
 		private readonly Audit _audit;
+		private readonly long? _currentElapsedTicks;
+
+		private bool _isDisposed;
+		private Stopwatch _sw;
```
Comment (russcam) [inline src/Elasticsearch.Net/Auditing/Auditable.cs]: `_sw` Not used?

**Comment by Mpdreamz** [comment]
The overal time is returned on this line

Audit trail of this API call (Total Elapsed: 00:00:06.3458422):

O
Overal structure:

```
[Overal time]
- [Audit overal time]
---[Profile Step time]
```

Will have to investigate if rooting it to `Transport` is feasible.

**Comment by russcam** [comment]
look at after 6.x release

**Comment by Mpdreamz** [comment]
We can finally close this now its re-implemented using `Activity/DiagnosticSource` :tada: 

see: #3862

**Pull request closed without merging**